lib/core/connection/pool.js

```diff
@@ -631,43 +654,30 @@ Pool.prototype.unref = function() {
   });
 };
 
-// Events
-var events = ['error', 'close', 'timeout', 'parseError', 'connect', 'message'];
-
 // Destroy the connections
 function destroy(self, connections, options, callback) {
-  let connectionCount = connections.length;
-  function connectionDestroyed() {
-    connectionCount--;
-    if (connectionCount > 0) {
-      return;
-    }
-
-    // Zero out all connections
-    self.inUseConnections = [];
-    self.availableConnections = [];
-    self.connectingConnections = 0;
+  eachAsync(
+    connections,
+    (conn, cb) => {
+      for (const eventName of CONNECTION_EVENTS) {
+        conn.removeAllListeners(eventName);
+      }
 
-    // Set state to destroyed
-    stateTransition(self, DESTROYED);
-    if (typeof callback === 'function') {
-      callback(null, null);
-    }
-  }
+      conn.destroy(options, cb);
+    },
+    err => {
+      if (err) {
+        if (typeof callback === 'function') callback(err, null);
+        return;
+      }
 
-  if (connectionCount === 0) {
-    connectionDestroyed();
-    return;
-  }
+      resetPoolState(self);
+      self.queue = [];
```

Comment:
Why don't we do this in `resetPoolState`?

Reply:
because when we reset the pool we don't want to clear the work queue, only when we destroy the pool. I could have made this a boolean parameter perhaps, but again we're not trying to get too tricky with our updates to this code which should go away soon